File: 2024/july/java/16.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean solve(TreeNode root , int target ,StringBuilder ans){
        if (root == null) return false;
        if (root.val == target){
            return true;
        }
        ans.append('L');
        boolean ch = solve(root.left , target, ans);
        if (ch){
            return true;
        }
        ans.deleteCharAt(ans.length() - 1);
        ans.append('R');
        ch = solve(root.right , target , ans);
        if (ch){
            return true;
        }
        ans.deleteCharAt(ans.length() - 1);
         
        return false;
    }
    public String getDirections(TreeNode root, int startValue, int destValue) {
        StringBuilder destPath = new StringBuilder();
        StringBuilder startPath = new StringBuilder();
        solve(root , startValue , startPath);
        solve(root , destValue , destPath);
        StringBuilder ans = new StringBuilder();
        int common = 0;
        int sz1 = destPath.length();
        int sz0 = startPath.length();
        while(common < sz0 && common < sz1 && startPath.charAt(common) == destPath.charAt(common) ) ++common;

        for (int i = 0; i < sz0 - common; ++i) ans.append('U');
        for (int i = common; i < sz1; ++i)ans.append(destPath.charAt(i));
        return ans.toString();
       
        
    }
}

File: 2024/july/javaScript/16.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} startValue
 * @param {number} destValue
 * @return {string}
 */
var getDirections = function(root, startValue, destValue) {

    function solve(root ,target , ans){
        if (root == null) return false;
        if (root.val == target){
            return true;
        }
        ans.push('L');
        let ch = solve(root.left , target, ans);
        if (ch){
            return true;
        }
        ans.pop();
        ans.push('R');
        ch = solve(root.right , target , ans);
        if (ch){
            return true;
        }
        ans.pop();
         
        return false;
    }
        let destPath = [];
        let startPath = [];
        solve(root , startValue , startPath);
        solve(root , destValue , destPath);
        let ans = "";
        let common = 0;
        const sz1 = destPath.length;
        const sz0 = startPath.length;
        while(common < sz0 && common < sz1 && startPath.at(common) == destPath.at(common) ) ++common;

        for (let i = 0; i < sz0 - common; ++i) ans+='U';
        for (let i = common; i < sz1; ++i)ans+=destPath.at(i);
        return ans;
    
};
